services/data_sources/json_notification_data_source.rb

```diff
@@ -0,0 +1,162 @@
+
+    class << self
+      attr_accessor :file_semaphore
+    end
+
+    # @return [FastlaneCI::GitRepo]
+    attr_accessor :git_repo
+
+    # Can't have us reading and writing to a file at the same time
+    JSONNotificationDataSource.file_semaphore = Mutex.new
+
+    # Reloads notifications from the notifications data source after instantiation
+    #
+    # @param  [Any] params
+    def after_creation(**params)
+      if params.nil?
+        raise "Either user or a provider credential is mandatory."
+      else
+        if !params[:user] && !params[:provider_credential]
+          raise "Either user or a provider credential is mandatory."
+        else
+          params[:provider_credential] ||= params[:user].provider_credential(type: ProviderCredential::PROVIDER_CREDENTIAL_TYPES[:github])
+          @git_repo = FastlaneCI::GitRepo.new(git_config: self.json_folder_path, provider_credential: params[:provider_credential])
+        end
+      end
+
+      reload_notifications
+    end
+
+    # Returns an array of notifications from the notifications JSON file stored
+    # in the configuration git repo
+    #
+    # @return [Array[Notification]]
+    def notifications
+      JSONProjectDataSource.projects_file_semaphore.synchronize do
+        path = git_repo.file_path("notifications.json")
+        return [] unless File.exist?(path)
+
+        return JSON.parse(File.read(path)).map(&Notification.method(:from_json!))
+      end
+    end
+
+    # Writes the notifications array to the git repo as JSON and commit them as
+    # the CI user
+    #
+    # @param  [Array[Notification]] notifications
+    def notifications=(notifications)
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        File.write(notifications_file_path, JSON.pretty_generate(notifications.map(&:to_object_dictionary)))
+        git_repo.commit_changes!
+      end
+    end
+
+    # Returns `true` if the notification exists in the in-memory notifications object
+    #
+    # @param  [String] id
+    # @return [Boolean]
+    def notification_exist?(id: nil)
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        existing_notification = @notifications.select { |notification| notification.id == id }.first
+        return existing_notification.nil? ? false : true
+      end
+    end
+
+    # Swaps the old notification record with the updated notification record if
+    # the notification exists
+    #
+    # @param  [Notification] notification
+    def update_notification!(notification: nil)
+      notification.updated_at = Time.now
+
+      notification_index = nil
+      existing_notification = nil
+
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        @notifications.each.with_index do |old_notification, index|
+          if old_notification.id == notification.id
+            notification_index = index
+            existing_notification = old_notification
+            break
+          end
+        end
+      end
```

Comment:
Not clear why this semaphore exists here. It doesn't look like it's wrapping any IO

Reply:
Oops, that was my bad 😳